In [13]:
#improting important libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
data_set = "WineQT.csv"

In [10]:
#converting csv file into dataframe
Data_frame = pd.read_csv(data_set)

#data exploration steps
print(Data_frame.head(5))
Data_frame.info()

   fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  Id  
0      9.4        5   0  
1      9.8        5   1  
2      9

In [12]:
# Split the data into features X and target variable y
feature_set = Data_frame.drop('quality', axis=1)
target_class = Data_frame['quality']

#split training and testing set
training_features, testing_features, training_label, testing_label = train_test_split(feature_set, target_class, train_size=0.75, random_state=42)

#normalizing the features using StandardScaler
scaler = StandardScaler()
scaled_trainfeatures = scaler.fit_transform(training_features)
scaled_testfeatures = scaler.transform(testing_features)

We are going to assess various regression models to predict wine quality, comparing their performance using Root Mean Squared Error (RMSE). After preparing and splitting the dataset, models like Linear Regression, Ridge Regression, Lasso Regression, and Random Forest Regression to be trained and evaluated below, with the most accurate one selected based on RMSE.




In [16]:
# Initialize different regression models
linear_reg = LinearRegression()
ridge_reg = Ridge(alpha=1.0)
lasso_reg = Lasso(alpha=1.0)
random_forest_reg = RandomForestRegressor(n_estimators=100)

# Train each model on the scaled training features and corresponding labels
linear_reg.fit(scaled_trainfeatures, training_label)
ridge_reg.fit(scaled_trainfeatures, training_label)
lasso_reg.fit(scaled_trainfeatures, training_label)
random_forest_reg.fit(scaled_trainfeatures, training_label)

# Make predictions on the scaled testing features
linear_reg_pred = linear_reg.predict(scaled_testfeatures)
ridge_reg_pred = ridge_reg.predict(scaled_testfeatures)
lasso_reg_pred = lasso_reg.predict(scaled_testfeatures)
random_forest_reg_pred = random_forest_reg.predict(scaled_testfeatures)

# Evaluate each model's performance using mean squared error (MSE)
linear_reg_mse = mean_squared_error(testing_label, linear_reg_pred)
ridge_reg_mse = mean_squared_error(testing_label, ridge_reg_pred)
lasso_reg_mse = mean_squared_error(testing_label, lasso_reg_pred)
random_forest_reg_mse = mean_squared_error(testing_label, random_forest_reg_pred)

# Calculate RMSE for each model
linear_reg_rmse = np.sqrt(mean_squared_error(testing_label, linear_reg_pred))
ridge_reg_rmse = np.sqrt(mean_squared_error(testing_label, ridge_reg_pred))
lasso_reg_rmse = np.sqrt(mean_squared_error(testing_label, lasso_reg_pred))
random_forest_reg_rmse = np.sqrt(mean_squared_error(testing_label, random_forest_reg_pred))

print("Linear Regression RMSE:", linear_reg_rmse)
print("Ridge Regression RMSE:", ridge_reg_rmse)
print("Lasso Regression RMSE:", lasso_reg_rmse)
print("Random Forest Regression RMSE:", random_forest_reg_rmse)

Linear Regression RMSE: 0.6129365427029398
Ridge Regression RMSE: 0.6128394083356609
Lasso Regression RMSE: 0.7561017015609562
Random Forest Regression RMSE: 0.5634657986903077


We observe that the Random Forest Regression model has the lowest RMSE,which is 056, indicating good performance in predicting wine quality compared to the other models. Therefore, the Random Forest Regression model may be considered the most suitable for this regression task. Now we are going to apply classification models including randomforest and will comapre them.

In [18]:
# Encode wine quality into classes: low, medium, high
label_encoder = LabelEncoder()
wine_quality_classes = ['low', 'medium', 'high']
quality_labels = label_encoder.fit_transform(pd.cut(Data_frame['quality'], bins=[0, 4, 6, 10], labels=wine_quality_classes, right=False))

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(feature_set, quality_labels, train_size=0.75, random_state=42)

# Train the Random Forest Classifier
random_forest_classifier = RandomForestClassifier(n_estimators=100)
random_forest_classifier.fit(X_train, y_train)

# Make predictions on the test set
y_pred = random_forest_classifier.predict(X_test)

# Evaluate the performance of the model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)


Accuracy: 0.7832167832167832


In [19]:
#Logistic Regression model
logistic_reg = LogisticRegression()

# Train the Logistic Regression model on the scaled training features and corresponding labels
logistic_reg.fit(scaled_trainfeatures, training_label)

# Make predictions on the scaled testing features
logistic_reg_pred = logistic_reg.predict(scaled_testfeatures)

# Evaluate the Logistic Regression model's performance using accuracy
logistic_reg_accuracy = accuracy_score(testing_label, logistic_reg_pred)

print("Logistic Regression Accuracy:", logistic_reg_accuracy)

Logistic Regression Accuracy: 0.6363636363636364


Among both Logistic regression and Random forest models, Random forest classifier gave the higher accuracy which is 0.78.